### ISSUE TITLE
Question about writing a customized model with the loop function

### ISSUE THREAD

**Issue by ajing** [topepo/caret#667]
I have a question about writing customized methods when there is a loop function.

Here, I basically add one more variable _gamma_ to glmnet training. What should I change in the loop function? Currently, I just add gamma to the loop attribute. Is this the right way to do it?

It seems to work after I change the loop function.

```
glmnetCostSens <- list(label = "glmnetWeights",
                  library = c("glmnet", "Matrix"),
                  type = c("Regression", "Classification"),
                  parameters = data.frame(parameter = c('alpha', 'lambda', 'gamma'),
                                          class = c("numeric", "numeric", "numeric"),
                                          label = c('Mixing Percentage', 'Regularization Parameter', 'Class Weight')),
                  grid = function(x, y, len = NULL, search = "grid") {
                    if(search == "grid") {
                      numLev <- if(is.character(y) | is.factor(y)) length(levels(y)) else NA
                      if(!is.na(numLev)) {
                        fam <- ifelse(numLev > 2, "multinomial", "binomial")
                      } else fam <- "gaussian"
                      init <- glmnet(as.matrix(x), y,
                                     family = fam,
                                     nlambda = len+2,
                                     alpha = .5)
                      lambda <- unique(init$lambda)
                      lambda <- lambda[-c(1, length(lambda))]
                      lambda <- lambda[1:min(length(lambda), len)]

                      gamma_v = seq(0.05, 0.95, length.out = len)
                      gamma_v = unique(sort(c(gamma_v, 0.5)))

                      out <- expand.grid(alpha = seq(0.1, 1, length = len),
                                         lambda = lambda,
                                         gamma = gamma_v)
                    } else {
                      gamma_v = runif(len, min = 1, max = 25)
                      gamma_v = unique(sort(c(gamma_v, 0.5)))

                      out <- data.frame(alpha = runif(len, min = 0, 1),
                                        lambda = 2^runif(len, min = -10, 3),
                                        gamma = gamma_v)
                    }
                    out
                  },
                  loop = function(grid) {
                    alph_gamma <- unique(grid[, c("alpha", "gamma")])
                    loop <- alph_gamma
                    loop$lambda <- NA
                    submodels <- vector(mode = "list", length = nrow(alph_gamma))
                    for(i in seq(nrow(alph_gamma))) {
                      np <- merge(grid, alph_gamma[i, ], by = c("alpha", "gamma"))[, "lambda"]
                      loop$lambda[loop$alpha == alph_gamma$alpha[i] & loop$gamma == alph_gamma$gamma[i]] <- np[which.max(np)]
                      submodels[[i]] <- expand.grid(lambda = np[-which.max(np)])
                    }
                    #print(loop)
                    list(loop = loop, submodels = submodels)
                  },
                  fit = function(x, y, wts, param, lev, last, classProbs, ...) {
                    numLev <- if(is.character(y) | is.factor(y)) length(levels(y)) else NA

                    theDots <- list(...)

                    if(all(names(theDots) != "family")) {
                      if(!is.na(numLev)) {
                        fam <- ifelse(numLev > 2, "multinomial", "binomial")
                      } else fam <- "gaussian"
                      theDots$family <- fam
                    }

                    ## pass in any model weights
                    #print(names(param))
                    #print(param$gamma)
                    vWeight <- rep(1 - param$gamma, length(y))
                    ## assume the second in the lev is the positive label
                    #message(paste0("[glmnetCostSens] The possitve label is ", lev[2]))
                    vWeight[y == lev[2]] <- param$gamma

                    #print(vWeight)

                    theDots$weights <- vWeight

                    if(!(class(x)[1] %in% c("matrix", "sparseMatrix")))
                      x <- as.matrix(x)

                    modelArgs <- c(list(x = x,
                                        y = y,
                                        alpha = param$alpha),
                                   theDots)

                    out <- do.call("glmnet", modelArgs)
                    if(!is.na(param$lambda[1])) out$lambdaOpt <- param$lambda[1]
                    out
                  },
                  predict = function(modelFit, newdata, submodels = NULL) {
                    if(!is.matrix(newdata)) newdata <- as.matrix(newdata)
                    if(length(modelFit$obsLevels) < 2) {
                      out <- predict(modelFit, newdata, s = modelFit$lambdaOpt)
                    } else {
                      out <- predict(modelFit, newdata, s = modelFit$lambdaOpt, type = "class")
                    }
                    if(is.matrix(out)) out <- out[,1]

                    if(!is.null(submodels)) {
                      if(length(modelFit$obsLevels) < 2) {
                        tmp <- as.list(as.data.frame(predict(modelFit, newdata, s = submodels$lambda)))
                      } else {
                        tmp <- predict(modelFit, newdata, s = submodels$lambda, type = "class")
                        tmp <- if(is.matrix(tmp)) as.data.frame(tmp, stringsAsFactors = FALSE) else as.character(tmp)
                        tmp <- as.list(tmp)
                      }
                      out <- c(list(out), tmp)
                    }
                    out
                  },
                  prob = function(modelFit, newdata, submodels = NULL) {
                    obsLevels <- if("classnames" %in% names(modelFit)) modelFit$classnames else NULL
                    probs <- predict(modelFit,
                                     as.matrix(newdata),
                                     s = modelFit$lambdaOpt,
                                     type = "response")
                    if(length(obsLevels) == 2) {
                      probs <- as.vector(probs)
                      probs <- as.data.frame(cbind(1-probs, probs))
                      colnames(probs) <- modelFit$obsLevels
                    } else {
                      probs <- as.data.frame(probs[,,1,drop = FALSE])
                      names(probs) <- modelFit$obsLevels
                    }
                    if(!is.null(submodels)) {
                      tmp <- predict(modelFit,
                                     as.matrix(newdata),
                                     s = submodels$lambda,
                                     type = "response")
                      if(length(obsLevels) == 2) {
                        tmp <- as.list(as.data.frame(tmp))
                        tmp <- lapply(tmp,
                                      function(x, lev) {
                                        x <- as.vector(x)
                                        tmp <- data.frame(1-x, x)
                                        names(tmp) <- lev
                                        tmp
                                      },
                                      lev = modelFit$obsLevels)
                      } else tmp <- apply(tmp, 3, function(x) data.frame(x))
                      probs <- if(is.list(tmp)) c(list(probs), tmp) else list(probs, tmp)
                    }
                    probs
                  },
                  predictors = function(x, lambda = NULL, ...) {
                    if(is.null(lambda))
                    {
                      if(length(lambda) > 1) stop("Only one value of lambda is allowed right now")
                      if(!is.null(x$lambdaOpt)) {
                        lambda <- x$lambdaOpt
                      } else stop("must supply a value of lambda")
                    }
                    allVar <- if(is.list(x$beta)) rownames(x$beta[[1]]) else rownames(x$beta)
                    out <- unlist(predict(x, s = lambda, type = "nonzero"))
                    out <- unique(out)
                    if(length(out) > 0) {
                      out <- out[!is.na(out)]
                      out <- allVar[out]
                    }
                    out
                  },
                  varImp = function(object, lambda = NULL, ...) {
                    if(is.null(lambda)) {
                      if(length(lambda) > 1) stop("Only one value of lambda is allowed right now")
                      if(!is.null(object$lambdaOpt)) {
                        lambda <- object$lambdaOpt
                      } else stop("must supply a value of lambda")
                    }
                    beta <- predict(object, s = lambda, type = "coef")
                    if(is.list(beta)) {
                      out <- do.call("cbind", lapply(beta, function(x) x[,1]))
                      out <- as.data.frame(out)
                    } else out <- data.frame(Overall = beta[,1])
                    out <- abs(out[rownames(out) != "(Intercept)",,drop = FALSE])
                    out
                  },
                  levels = function(x) if(any(names(x) == "obsLevels")) x$obsLevels else NULL,
                  tags = c("Generalized Linear Model", "Implicit Feature Selection",
                           "L1 Regularization", "L2 Regularization", "Linear Classifier",
                           "Linear Regression"),
                  sort = function(x) x[order(-x$lambda, x$alpha, abs(x$gamma - 0.5)),],
                  trim = function(x) {
                    x$call <- NULL
                    x$df <- NULL
                    x$dev.ratio <- NULL
                    x
                  })


```

**Comment by topepo** [comment]
This is long just so I can outline the thought process. See tl;dr at the bottom if you just want the answer. 

You should only have to change `grid`, `loop` and `fit`. 

For `loop`, the goal is to get a core set of models that should be fit (the `loop` part of the list) and a list of submodels (one for every loop row) that has the values of `alpha` and `gamma`. 

`glmnet` is a bit odd compared to other models; most would have you choose the largest model possible for the `loop` specifications. For example, if a classification tree, you fit the deepest model in the `loop` and derive predictions from smaller models. `glmnet` does everything at once so there isn't a concept of the "full model" in this same way. It does need a value though, so it selects the largest value (see `getModelInfo("glmnet", FALSE)[[1]]$loop`):
```r
> glmn_grid <- expand.grid(alpha = c(.01, .1, 1),
+                          lambda = 2^(-2:0))
> 
> getModelInfo("glmnet", FALSE)[[1]]$loop(glmn_grid)
$loop
  alpha lambda
1  0.01      1
2  0.10      1
3  1.00      1

$submodels
$submodels[[1]]
  lambda
1   0.25
2   0.50

$submodels[[2]]
  lambda
1   0.25
2   0.50

$submodels[[3]]
  lambda
1   0.25
2   0.50
```

You would only need to add `gamma` to the `loop` part of the output by getting the unique combinations of `alpha` and `gamma`. 

**tldr**

Here you go
```r
## This can be used as the `loop` module
looper <- function (grid)  {
  loop <- plyr::ddply(grid, c("alpha", "gamma"),
                      function(x) c(lambda = max(x$lambda)))
  submodels <- vector(mode = "list", length = nrow(loop))
  for (i in seq(along = loop$lambda)) {
    index <- which(grid$gamma == loop$gamma[i] &
                   grid$alpha == loop$alpha[i])
    this_combo <- grid[index, "lambda"]
    submodels[[i]] <-
      data.frame(lambda = this_combo[this_combo != loop$lambda[i]])
  }
  list(loop = loop, submodels = submodels)
}
```
```r
> looper(glmn_grid)
$loop
  alpha gamma lambda
1  0.01  0.25      1
2  0.01  0.50      1
3  0.01  0.75      1
4  0.10  0.25      1
5  0.10  0.50      1
6  0.10  0.75      1
7  1.00  0.25      1
8  1.00  0.50      1
9  1.00  0.75      1

$submodels
$submodels[[1]]
  lambda
1   0.25
2   0.50

$submodels[[2]]
  lambda
1   0.25
2   0.50

$submodels[[3]]
  lambda
1   0.25
2   0.50

$submodels[[4]]
  lambda
1   0.25
2   0.50

$submodels[[5]]
  lambda
1   0.25
2   0.50

$submodels[[6]]
  lambda
1   0.25
2   0.50

$submodels[[7]]
  lambda
1   0.25
2   0.50

$submodels[[8]]
  lambda
1   0.25
2   0.50

$submodels[[9]]
  lambda
1   0.25
2   0.50
```

**Comment by topepo** [comment]
You might also want to use 
```r
theDots$weights <- 
  ifelse(y == lev[2], param$gamma, 1- param$gamma)
```

Let me know how this works out. I played around with this in regular logistic regression but all that seems to do is to change the intercept value. There is a paper on why that is (from SAS maybe) circulating somewhere on this topic. I would be interested to see if it is different for the `glmnet`.

**Comment by ajing** [comment]
Yes, it works. For my case, I need to optimize specificity with a minimum sensitivity and changing weights seems to work in some situations.
